File: 79-word-search/79-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        self.dirs = [(1,0),(-1,0),(0,1),(0,-1)]
        
        for i in range(len(board)):
            for j in range(len(board[0])):
                if self.dfs(i, j, board, word):
                    return True
        return False
        
    def dfs(self, x, y, board, word):
        if len(word) == 0:
            return True
        if 0 <= x < len(board) and 0 <= y < len(board[0]) and board[x][y] == word[0]:
            tmp = board[x][y]
            board[x][y] = "!"
            for dx, dy in self.dirs:
                newx, newy = x + dx, y + dy
                if self.dfs(newx, newy, board, word[1:]):
                    return True
            board[x][y] = tmp
        return False
    
    
                
        

File: 79-word-search/79-word-search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        char[] w = word.toCharArray();
        for (int x = 0; x < board.length; x++){
            for (int y = 0; y < board[0].length; y++){
                if (dfs(x, y, w, board, 0)) {
                    return true;
                }   
            }   
        }
        return false;
    }
    
    private boolean dfs(int x, int y, char[] w, char[][] board, int i){
        if (i == w.length) {
            return true;
        }
        
        if ( x < 0 || y < 0 || x == board.length || y == board[0].length || board[x][y] != w[i]){
            return false;
        } 
        
        char tmp = board[x][y];
        board[x][y] = '!';
        boolean trueExist = 
            dfs(x+1, y, w, board, i+1) ||
            dfs(x-1, y, w, board, i+1) ||
            dfs(x, y+1, w, board, i+1) ||
            dfs(x, y-1, w, board, i+1);
        board[x][y] = tmp;
        return trueExist;
    }
}